README.md

```diff
@@ -0,0 +1,116 @@
+# Devise: Setup and customization
+
+Set up a Rails server with [Devise] to let users log in with a username and password.
+
+## Objectives
+
+We're going to use Devise and Rails to create an app with a complete signin flow. We will basically be able to do this using only generators.
+
+## Instructions: Part 1, setup
+
+Add Devise to your Gemfile:
+
+    gem 'devise'
+
+Now run the installer:
+
+    rails generate devise:install
+
+This creates a massive initializer in `config/initializers/devise.rb`. This is probably the single best source of documentation for Devise that I've seen. You may wish to look through it.
+
+You'll notice that the installer prints a big notice of several things you should do. In particular, we should have a route route.
+
+Create a `WelcomeController` with a `home` action and a view that just prints `current_user.email`.
+
+Now generate your `User` model with:
+
+    rails g devise User
+
+Run `rake routes` and `rake db:migrate`. You should see that Devise has added a bunch of them. Run `rails s` and take a look at one, maybe `/users/sign_in`.
+
+You should now have a working app with sign in.
+
+## Part 2, customization
+
+Now let's add Facebook login support with Omniauth.
+
+Add omniauth-facebook to your Gemfile:
+
+    gem 'omniauth-facebook'
+
+We'll also need to store two more columns in our user model: a `provider` string (which will always be `'facebook'` or `nil` in our app for the moment), and a `uid`, the user's authenticated Facebook ID.
+
+    rails g migration AddOmniauthToUsers provider:index uid:index
+    rake db:migrate
+
+And add the Omniauth configuration to `config/initializers/devise.rb`:
+
+    config.omniauth :facebook, ENV['FACEBOOK_KEY'], ENV['FACEBOOK_SECRET']
+
+You'll see that there's another similar line commented out for Github; feel free to reference it. This requires that you have `FACEBOOK_KEY` and `FACEBOOK_SECRET` in your environment. Create an application in the [Facebook developer console][fbdev] and get these values from there. You'll also need to go to Settings > Advanced and set Valid OAuth Redirect URLs to include `http://localhost:3000/users/auth/facebook/callback`. This setting is listed under `Client OAuth Settings` in the dashboard, and yes, I wish I could link you right to it.
+
```

Comment:
i've seen students get really stuck on this stuff.  if we could record a little video, or do screenshots i think that would be enormously helpful.